kernel/eventfddev.c

```diff
@@ -0,0 +1,537 @@
+
+    if (ret > 0)
+        myst_tcall_poll_wake();
+
+    return ret;
+}
+
+static ssize_t _write(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    const void* buf,
+    size_t count)
+{
+    ssize_t ret = 0;
+    uint64_t value;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EBADF);
+
+    if (!buf)
+        ERAISE(-EINVAL);
+
+    if (count < sizeof(uint64_t))
+        ERAISE(-EINVAL);
+
+    memcpy(&value, buf, sizeof(uint64_t));
+
+    if (value == 0xffffffffffffffff)
+    {
+        _unlock(eventfd);
+        ERAISE(-EINVAL);
+    }
+
+    _lock(eventfd);
+
+    if (!_valid_eventfd(eventfd))
+    {
+        /* cannot unlock since mutex is no longer valid */
+        ERAISE(-EBADF);
+    }
+
+    /* if able to write the the counter */
+    if (eventfd->counter + value < MAX_COUNTER_VALUE)
+    {
+        eventfd->counter += value;
+        myst_cond_signal(&eventfd->cond);
+        _unlock(eventfd);
+        ret = sizeof(uint64_t);
+        goto done;
+    }
+
+    /* handle non-blocking case */
+    if (eventfd->flags & EFD_NONBLOCK)
+    {
+        _unlock(eventfd);
+        ERAISE(-EAGAIN);
+    }
+
+    /* wait here for another thread to read the counter */
+    for (;;)
+    {
+        if (myst_cond_wait(&eventfd->cond, &eventfd->mutex) != 0)
+        {
+            /* unexpected */
+            _unlock(eventfd);
+            ERAISE(-EPIPE);
+        }
+
+        if (eventfd->counter + value < MAX_COUNTER_VALUE)
+        {
+            eventfd->counter += value;
+            myst_cond_signal(&eventfd->cond);
+            _unlock(eventfd);
+            ret = sizeof(uint64_t);
+            goto done;
+        }
+    }
+
+    _unlock(eventfd);
+    ret = count;
+
+done:
+
+    if (ret > 0)
+        myst_tcall_poll_wake();
+
+    return ret;
+}
+
+static ssize_t _readv(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    const struct iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = 0;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EINVAL);
+
+    ret = myst_fdops_readv(&eventfddev->fdops, eventfd, iov, iovcnt);
+    ECHECK(ret);
+
+done:
+
+    return ret;
+}
+
+static ssize_t _writev(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    const struct iovec* iov,
+    int iovcnt)
+{
+    ssize_t ret = 0;
+
+    if (!eventfddev || !_valid_eventfd(eventfd))
+        ERAISE(-EINVAL);
+
+    ret = myst_fdops_writev(&eventfddev->fdops, eventfd, iov, iovcnt);
+    ECHECK(ret);
+
+done:
+
+    return ret;
+}
+
+static int _fstat(
+    myst_eventfddev_t* eventfddev,
+    myst_eventfd_t* eventfd,
+    struct stat* statbuf)
+{
+    int ret = 0;
+    struct stat buf;
+
+    if (!eventfddev || !_valid_eventfd(eventfd) || !statbuf)
+        ERAISE(-EINVAL);
+
+    memset(&buf, 0, sizeof(buf));
+    buf.st_dev = 13; /* event fd */
+    buf.st_ino = (ino_t)eventfd;
+    buf.st_mode = 0600;
+    buf.st_nlink = 1;
+    buf.st_uid = MYST_DEFAULT_UID;
```

Comment:
We probably should use `thread->uid`. But I doubt it makes any difference to any application.

Reply:
Good catch! I verified that uid and gid for eventfd's is 0, so I changed both to zero.